app/src/main/java/it/niedermann/owncloud/notes/persistence/NotesDatabase.java

```diff
@@ -769,4 +842,102 @@ private static void validateAccountId(long accountId) {
             throw new IllegalArgumentException("accountId must be greater than 0");
         }
     }
+
+    /**
+     * Get the category if with the given category title
+     * The method does not support fuzzy search.
+     * Because the category title in database is unique, there will not at most one result.
+     * If there is no such category, database will create it if create flag is set.
+     * Otherwise this method will return -1 as default value.
+     *
+     * @param accountId     The user account Id
+     * @param categoryTitle The category title which will be search in the db
+     * @param create        True if there is no such category, the db will add one.
+     *                      False the db will not create one and return -1
+     * @return -1 if there is no such category else the corresponding id
+     */
+    @NonNull
+    @WorkerThread
+    private Integer getCategoryIdByTitle(long accountId, @NonNull String categoryTitle, boolean create) {
+        if (create) {
+            // If the category does not exist in the current db, create it.
+            if (getCategoryIdByTitle(accountId, categoryTitle, false) == -1 &&
+                    addCategory(categoryTitle, accountId) == -1) {
+                Log.e(TAG, String.format("Error occurs when creating category: %s", categoryTitle));
+            }
+        }
+        validateAccountId(accountId);
+        SQLiteDatabase db = getReadableDatabase();
+        Cursor cursor = db.query(
+                table_category,
+                new String[]{key_id},
+                key_title + " = ? AND " + key_account_id + " = ? ",
+                new String[]{categoryTitle, String.valueOf(accountId)},
+                key_id,
+                null,
+                key_id);
+        int id = -1;
+        if (cursor.moveToNext()) {
+            id = cursor.getInt(0);
+        }
+        cursor.close();
+        return id;
+    }
```

Comment:
Because I found that the return value of `addCategory()` is the row ID of the newly inserted row. But the row ID I think is sth different from the primary key ID? I don't know if I got it wrong but if you can make sure that the return value of `addCategory()` is exactly the primary key ID,  I will do it.

Reply:
According to the answers on Stackoverflow, it is the same. But we will see if it's working or not ...